# Functions Assignments

## Define a function named addition that takes two parameters a and b and returns their sum.

- Input: addition(3, 5)
- Output: 8

In [1]:
def addition(a,b):
    return a+b

addition(3,5)

8

## Write a function called multiply that accepts two arguments and returns the result of multiplying them together.

- Input: multiply(4, 6)
- Output: 24

In [2]:
def multiply(a, b):
    return a*b
multiply(3,3)

9

In [29]:
def multiply(a,b):
    c = a*b
    return c

multiply(3,5)

15

## Create a function named power that takes two parameters: base and exponent, and returns base raised to the power of exponent.

- Input: power(2, 3)
- Output: 8

In [30]:
def power(a,b):
    c = a**b
    return c

power(3,2)

9

In [4]:
def is_even(num):
    if num % 2 == 0:
        return True
    else:
        return False
    
is_even(9)

False

In [ ]:
def is_even(num):
    return num % 2 == 0

## Define a function named is_even that takes an integer parameter num and returns True if the number is even, otherwise False.

- Input: is_even(7)
- Output: False

In [31]:
# Write your code here
def is_even(num):
    if num % 2 == 0:
        return True
    else:
        return False
    
is_even(8)

True

In [ ]:
for i in range(5,0,-1):
    print(i)
    5 4 3 2 1 0

5
4
3
2
1


In [12]:
def factorial(a):
    total = 1
    for n in range(a,0,-1):
        total *= n
    return total

factorial(5)

120

## Write a function called factorial that calculates the factorial of a given number n.

- Input: factorial(5)
- Output: 120

In [32]:
# Write your code here
# 5 = 5*4*3*2*1
def factorial(n):
    for i in range(1,n):
        n *= i
    return n

factorial(3)

6

In [22]:
5*4*3*2*1

120

In [23]:
eval('*'.join([str(i) for i in range(5,0,-1)]))

120

In [ ]:
# List Comprehension
def factorial(num):
    [str(i) for n in num ]

'3'

## Create a function named is_palindrome that takes a string parameter text and returns True if the text is a palindrome, otherwise False.

- Input: is_palindrome("radar")
- Output: True

- Note: A palindrome is a word, phrase, number, or other sequence of characters that reads the same forward and backward.Example: "radar" is a palindrome because it reads the same forward and backward.

In [18]:
def is_palindeome(word):
    revrse_word=word[::-1]
    if word.lower()==revrse_word.lower():
        return True
    else:
        return False

is_palindeome("radar")

True

In [ ]:
def unique_elements(lst):
    result = []
    for n in lst:
        if n not in result:
            result.append(n)
    return result

# def unique_elements(lst):
#     return [n for n in lst if lst.count(n)  1]

unique_elements([1,2,2,3])

[]

In [25]:
def unique_elements(lst):
    result = set(lst)
    return list(result)

unique_elements([1,2,2,3])

[1, 2, 3]

## Define a function named unique_elements that takes a list parameter lst and returns a list contaiinng only the unique elements of the input list.
`
- Input: unique_elements([1, 2, 3, 3, 4, 5, 5])
- Output: [1, 2, 3, 4, 5]

In [20]:
def unique_elements(ist):
    unique_elements = []
    for i in ist:
        if i not in unique_elements:
            unique_elements.append(i)
    return unique_elements
unique_elements([1, 2, 3, 3, 4, 5, 5,6,6,7])

[1, 2, 3, 4, 5, 6, 7]

In [38]:
def average(lst):
    return sum(lst)/len(lst)

average([1,2,3])

def average(lst):
    avg = 0
    length = 0
    for i in lst:
        avg += i
        length += 1
    
    return avg/length

average([1,2,4])

2.3333333333333335

## Write a function called average that calculates the average of a list of numbers passed as a parameter.

- Input: average([1, 2, 3, 4, 5])
- Output: 3.0

In [26]:
def average(n):
#     print(n)
    return sum(n)/len(n)

average([1,2,3,4,5])

[1, 2, 3, 4, 5]


3.0

## Create a function named is_prime that takes an integer parameter num and returns True if the number is prime, otherwise False.

- Input: is_prime(7)
- Output: True

In [47]:
# Write your code here
def is_prime(n):
    if n < 2:
        return False
    i = 2
    while i*i <= n:
        if n % i == 0:
            return False
        i += 1
    return True
        
is_prime(6)


False

## Define a function named reverse_list that takes a list parameter lst and returns a new list with the elements reversed.

- Input: reverse_list([1, 2, 3, 4, 5])
- Output: [5, 4, 3, 2, 1]

In [49]:
def reverse_list(lst):
    return lst[::-1]

reverse_list([1,2,3])

[3, 2, 1]

In [39]:
def even(lst):
    return [n for n in lst if n%2 == 0]

even([1,2,3,4,5,6,7])

[2, 4, 6]

## Write a Python program to print the even numbers from a given list.

- Sample List : [1, 2, 3, 4, 5, 6, 7, 8, 9]
- Expected Result : [2, 4, 6, 8]

In [55]:
def even(lst):
    even = []
    odd = []
    for i in lst:
        if i % 2 == 0:
            even.append(i)
        else:
            odd.append(i)
    return even, odd

even([1, 2, 3, 4, 5, 6, 7, 8, 9])
    

([2, 4, 6, 8], [1, 3, 5, 7, 9])

In [54]:
def is_even_odd(lst):
    even = [i for i in lst if i % 2 == 0]
    odd = [i for i in lst if i % 2 != 0]
    print(f'''
    Even Numbers: {even}
    Odd Numbers: {odd}
    ''')

is_even_odd([1, 2, 3, 4, 5, 6, 7, 8, 9])



    Even Numbers: [2, 4, 6, 8]
    Odd Numbers: [1, 3, 5, 7, 9]
    


In [41]:
def count(string):
    lower = 0 
    upper = 0 
    for char in string:
        if char.islower():
            lower += 1
        elif char.isupper():
            upper += 1
    print(f'''
No. of Upper case characters : {upper}
No. of Lower case Characters : {lower}
'''   )

count('ali KHAN')


No. of Upper case characters : 4
No. of Lower case Characters : 3



## Write a Python function that accepts a string and counts the number of upper and lower case letters.
Sample String : 'The quick Brow Fox'
    
- Expected Output :
- No. of Upper case characters : 3
- No. of Lower case Characters : 12

In [57]:
# Write your code here
# The quick Brow Fox
def text(string):
    upper = []
    lower = []
    for word in string:
        if word.isupper():
            upper.append(word)
        elif word.islower():
            lower.append(word)
    print(f'''
No. of Upper case characters : {len(upper)}
No. of Lower case Characters : {len(lower)}
    ''')
    
text('The quick Brow Fox')



No. of Upper case characters : 3
No. of Lower case Characters : 12
    


In [58]:
# Write your code here
# The quick Brow Fox
def text(string):
    upper = 0
    lower = 0
    for word in string:
        if word.isupper():
            upper+=1
        elif word.islower():
            lower+=1
    print(f'''
No. of Upper case characters : {upper}
No. of Lower case Characters : {lower}
    ''')
    
text('The quick Brow Fox')



No. of Upper case characters : 3
No. of Lower case Characters : 12
    


In [59]:
def text(string):
    upper = len([word for word in string if word.isupper()])
    lower = len([word for word in string if word.islower()])
    print(f'''
No. of Upper case characters : {upper}
No. of Lower case Characters : {lower}
    ''')
    
    
text('The quick Brow Fox')



No. of Upper case characters : 3
No. of Lower case Characters : 12
    


In [1]:
def multiply(lst):
    result = 1
    for num in lst:
        result *= num

    return result
multiply([2,3,4])

24

## Write a Python function to multiply all the numbers in a list.

- Sample List : (8, 2, 3, -1, 7)
- Expected Output : -336

In [62]:
# Write your code here
def multiply(list1):
    x=1
    for i in list1:
        x*=i
    return x
multiply([8, 2, 3, -1, 7])

-336

In [4]:
'aaa'.count('a')

3

In [11]:
def remove_duplicates(string):
    result = ''
    for char in string:
        if char not in result:
            result += char
        elif char == ' ':
            result += ' '
    return result



remove_duplicates('I am Ali khan ')

'I am Ali khn '

## Write a Python program for a given string S which may contain lowercase and uppercase characters. The task is to remove all duplicate characters from the string and find the resultant string.

- Input: Str = HappyNewYear
- Output: HapyNewYr
- Explanation: After removing duplicate characters such as p, e, a, we have string as “HapyNewYr”.

In [66]:
# Write your code _here
def removingduplicate(strr):
    result =""
    for i in strr:
        if i not in result:
            
            result = result + i
    return result
removingduplicate("HappyNewYear")
    

'HapyNewYr'

In [ ]:
def square(lst):
    result = {}
    for num in lst:
        result[num] = num**2
    return result 

square([1,2,3,4])

{1: 1, 2: 4, 3: 9, 4: 16}

In [17]:
def swap(lst, a, b):
    a = a -1
    b = b -1
    lst[a], lst[b] = lst[b], lst[a]
    return lst
swap([1,2,3,4], 1,3)

[3, 2, 1, 4]

In [19]:
def swap(lst, a, b):
    a-=1
    b-=1
    temp = lst[a]
    lst[a] = lst[b]
    lst[b] = temp
    return lst
swap([3,1,5,7], 1,3)

[5, 1, 3, 7]

## Define a function named square_dict that takes a list of integers as input and returns a dictionary where each key-value pair consists of an element from the input list as the key and its square as the value.

- Input: [1, 2, 3, 4, 5]
- Output: {1: 1, 2: 4, 3: 9, 4: 16, 5: 25}
- Note: You can use dictionary comprehension for better practices 

In [69]:
def square_dict(lst):
    return {i:i*i for i in lst}

square_dict([1,2,3,4,5])



{1: 1, 2: 4, 3: 9, 4: 16, 5: 25}

In [70]:
def square_dict(lst):
    dic = {}
    for i in lst:
        dic[i] = i*i
    return dic
        
#     return {i:i*i for i in lst}

square_dict([1,2,3,4,5])



{1: 1, 2: 4, 3: 9, 4: 16, 5: 25}

## Given a list in Python and provided the positions of the elements, write a fucntion to swap the two elements in the list.

- Input : List = [1, 2, 3, 4, 5], pos1 = 2, pos2 = 5
- Output : [1, 5, 3, 4, 2]

In [3]:
# write you code here
def swap(lst, pos1, pos2):
    lst[pos1] , lst[pos2] = lst[pos2], lst[pos1]
    return lst

swap([1, 2, 3, 4, 5], 1,-1)
# print('hello')

[1, 5, 3, 4, 2]

In [22]:
def remove(string, target):
    result = ''
    for char in string:
        if char != target:
            result += char
    return result

def remove(string, target):
    return ''.join([char for char in string if char != target])

remove('elephant', 't')


'elephan'

## Write a python function to remove a letter "e" from the string "elephant".

- output = 'lphant' (all 'e' characters are removed)

In [5]:
# Write your code here
def remove_e(a):
    result = ""
    for i in a:
        if i !="e":
            result = result + i
    return result

remove_e("elephant")


'lphant'

In [13]:
def remove(string):
    return ''.join([i for i in string if i!='e'])

remove('elephant')

'lphant'

In [24]:
def remove(string):
    result = ''
    for char in string:
        if string.count(char) == 1:
            result += char
    return result

def remove(string):
    return ''.join([char for char in string if string.count(char) == 1])
remove('levels')

'vs'

## Write a python function to remove duplicate character from a string.

- Input: 'levels'
- output: 'vs' (l and e are removed since they are repeating characters in the string)

In [15]:
# Write your code here
def rem(word):
    word2=''
    for i in word:
        if word.count(i)==1:
            word2+=i
    return word2
rem('levels')

'vs'

In [ ]:
def rem(string):
    dic = {}
    for i in string:
        if i not in dic.keys():
            dic[i] = 

In [ ]:
{
    'l':2,
    'e':2,
    'v':1,
    's':1
}

In [20]:
def rem(string):
    return ''.join([i for i in string if string.count(i) == 1])

rem('levels')

'vs'

In [35]:
def rem(string):
    dic={}
    for i in string:
        if i in dic:
            dic[i]+=1
        else:
            dic[i]=1
    output=''
    for i in string:
        if dic[i]==1:
            output+=i
    return output
rem("elephant")


'lphant'

In [38]:
def rem(string):
    dic={}
    for i in string:
        if i in dic:
            dic[i]+=1
        else:
            dic[i]=1
    output=''
    for key, value in dic.items():
        if value == 1:
            output+=key
    return output

rem("levels")


'vs'

In [30]:
'a' in {'b':'a'}

False

In [27]:
def count(string):
    result = {}
    for char in string:
        if char in result:
            result[char] += 1
        else:
            result[char] = 1
    return result
count('aali')


{'a': 2, 'l': 1, 'i': 1}

## Write a python function to count the frequencies of each character in the given string and stored it in a dictionary format.

- Input: 'listen'
- Output : {'l':1, 'i':1, 's':1, 't':1, 'e':1, 'n':1}

In [40]:
# Write your code here
def counting(string):
    dic = {}
    for i in string:
        if i in dic:
            dic[i]+=1
        else:
            dic[i] = 1
            
    return dic

counting('listen')

{'l': 1, 'i': 1, 's': 1, 't': 1, 'e': 1, 'n': 1}

In [34]:
def azfar(name):
    words = name.split()
    initials = '.'.join(word[0] for word in words)
    return initials
azfar('Vel Gopal Ayyer')

'V.G.A'

## Write a Python function named get_initials that takes a string as input and returns the initials of each word in the string, separated by dots.

- Input: "John Doe"
- Output: "J.D"

In [47]:
def get_initials(string):
    splitting=string.split()
    initial=""
    for i in splitting:
        initial+=i[0]
    dots_seperate=".".join(initial)    
    return dots_seperate


get_initials("Abdul Rafay")

'A.R'

In [57]:
def get_initials(string):
    a = string.split()
    
    c = a[0]
    d = a[1]
    return f"{c[0]}.{d[0]}"
get_initials("AMIN SADRUDDIN")
get_initials("KARIM SADRDDRUN")

'K.S'

In [65]:
def get_initials(string):
    result = ""
    for i in string.split():
        result =result + i[0] + "."
    return result[:-1]


get_initials("amin sadruddin roghi")


'a.s.r.'

In [71]:
def get_initials(string):
    return '.'.join([word[0].upper() for word in string.split()])

get_initials('waseem sattar jawad')

'W.S.J'